### PR TITLE
natural sort for filenames

### PR DESCRIPTION
**Pull request by jayay**
consider you have an album with following titles:

- track 1.wav
- track 2.wav
- track 3.wav
- ...
- track 10.wav
- track 11.wav

cmus doesn't order them 'the natural way'. Here's a new order function to fix this!

I've tested it on Fedora and Gentoo and it works well there. Please let me know if there are any issues!

### CODE REVIEW

**Comment by flyingmutant** [comment]
Thanks a lot for the PR! I've reviewed it briefly, and it looks OK in general. I'll try to find time for proper review this week.

**Comment by jayay** [comment]
Okay, and thanks for the reply. Please let me also know if you want the commits to be rebased.

File: track_info.c [hunk 1]
```diff
@@ -183,6 +186,55 @@ static int doublecmp0(double a, double b)
 	return (x > 0) - (x < 0);
 }
 
+int strnatcmp(const char * str1, const char * str2)
+{
+	int result;
+	char *o_ptr1, *o_ptr2, *strcpy1, *strcpy2;
+
+	o_ptr1 = xstrndup(str1, strlen(str1) + 1);
+	o_ptr2 = xstrndup(str2, strlen(str2) + 1);
```
Comment (flyingmutant) [inline track_info.c]: I don't see any real need to have `xstrndup`/`free` calls in `strnatcmp`. This is a comparator, and it must be fast (including not allocating memory).

File: track_info.c [hunk 2]
```diff
@@ -183,6 +186,55 @@ static int doublecmp0(double a, double b)
...
+	while (*strcpy1 != '\0' && isspace(*strcpy1))
+		strcpy1++;
+
+	while (*strcpy2 != '\0' && isspace(*strcpy2))
+		strcpy2++;
+
+	unsigned int pos = 0;
+	int num1 = 0, num2 = 0;
+
+	while ((unsigned int)(strcpy1 - o_ptr1 + pos) < strlen(strcpy1) &&
```
Comment (flyingmutant) [inline track_info.c]: `strlen` in a loop is a bad idea.

File: track_info.c [hunk 3]
```diff
@@ -183,6 +186,55 @@ static int doublecmp0(double a, double b)
...
+	       (unsigned int)(strcpy2 - o_ptr2 + pos) < strlen(strcpy2) &&
+		!isdigit(*(strcpy1+pos)) && !isdigit(*(strcpy2+pos)))
+		pos++;
+
+	result = strcoll(strcpy1, strcpy2);
+
+	if (pos) {
+		int cmp = strncmp(strcpy1, strcpy2, pos);
+
+		if (0 == cmp) {
```
Comment (flyingmutant) [inline track_info.c]: Please use `cmp == 0` instead.

**Comment by flyingmutant** [comment]
Sorry, I somehow missed the update to this PR — maybe GitHub is not sending emails when new commits are pushed to the PR branch?

File: track_info.c [hunk 4]
```diff
@@ -183,6 +186,52 @@ static int doublecmp0(double a, double b)
...
+	int num1 = 0, num2 = 0;
+	size_t strlen1 = strlen(strcpy1);
+	size_t strlen2 = strlen(strcpy2);
+
+	while ((size_t)(strcpy1 - str1 + pos) < strlen1 &&
+	       (size_t)(strcpy2 - str2 + pos) < strlen2 &&
+		!isdigit(*(strcpy1+pos)) && !isdigit(*(strcpy2+pos)))
+		pos++;
+
+	result = strcoll(strcpy1, strcpy2);
```
Comment (flyingmutant) [inline track_info.c]: You are computing `result` unconditionally here, and do not use it in `if (pos)` case.

File: track_info.c [hunk 5]
```diff
@@ -183,6 +186,52 @@ static int doublecmp0(double a, double b)
...
+
+	if (pos) {
```
Comment (flyingmutant) [inline track_info.c]: `if (pos)` branch will no be taken if the filenames consist entirely of digits. `strcoll` will be incorrectly used for this instead of numeric comparison.

**Comment by jayay** [comment]
Thanks for reviewing! I'll test some cases at the weekend and try to send another PR then.

**Comment by jayay** [comment]
Sorry, I still need some time because I'm messing around with a segfault appearing under circumstances I cannot figure out yet. :-1:

**Comment by flyingmutant** [comment]
Not a problem at all! It will be done when it will be done, no hurry.

**Comment by jayay** [comment]
It has become more complex, so i've put it into its own file.

File: track_info.c [hunk 6]
```diff
@@ -25,9 +25,13 @@
 #include "utils.h"
 #include "debug.h"
 #include "path.h"
+#include "strnatcmp.h"
 
 #include <string.h>
 #include <math.h>
+#include <ctype.h>
```
Comment (jayay) [inline track_info.c]: Oops! Forgot to remove these, I'll remove that in the next commit

Comment (jayay) [inline track_info.c]: same applies to line 33.

**Comment by flyingmutant** [comment]
Hi @jayay, I am waiting for your «next commit» — have you forgot to push it?

**Comment by jayay** [comment]
Okay @flyingmutant, that should be everything for now.

File: utils.h [hunk 7]
```diff
@@ -109,7 +110,7 @@ static inline int strcmp0(const char *str1, const char *str2)
 	if (!str2)
 		return 1;
 
-	return strcmp(str1, str2);
+	return strnatcmp(str1, str2);
```
Comment (flyingmutant) [inline utils.h]: Why this change?

Comment (flyingmutant) [inline utils.h]: OK, I see — https://github.com/jayay/cmus/commit/2cc7836f069a0a2e85db688e4b9c77bb78fea552.

However, please create a separate `strnatcmp0` and use that instead. `strcmp0` is also used in `lib.c`.

**Comment by jayay** [comment]
I wasn't sure there.

**Comment by flyingmutant** [comment]
Sorry for long delay! I've finally managed to review the PR properly.

I have a question — isn't https://github.com/flyingmutant/cmus/commit/cf5f6d20ccaafeddd86f8e4b6a218b51cd1f5eb0 the same thing, but with much less code? Or have I missed something (which is very likely)?

**Comment by jayay** [comment]
@flyingmutant Nice! That's a lot more readable! But I get a segmentation fault, whenever the file name contains braces.

**Comment by flyingmutant** [comment]
Can you please post the filenames and a backtrace? I can't debug the code myself right now, unfortunately.

**Comment by jayay** [comment]
Yes, of course!

File name: `02 - Kids (2 Finger an den Kopf).flac`
Output by GDB: 
```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe28e3700 (LWP 24101)]
0x000000000042a641 in skip_ws (str=str@entry=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t")
    at strnatcmp.c:27flac
        while (*str != '\0' && (isspace(*str) || iscntrl(*str) || !isprint(*str)))
 ```

backtrace:

```
#0  0x000000000042a641 in skip_ws (str=str@entry=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t")
    at strnatcmp.c:27üht.flac
#1  0x000000000042a6b8 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:42
#2  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#3  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#4  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#5  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#6  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#7  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#8  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#9  0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#10 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#11 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "",
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#12 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#13 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#14 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#15 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
#16 0x000000000042a7a9 in strnatcmp (str1=0x7fffd40014ea "", 
    str2=0x7fffd40044f1 "!\020\035\027\f\036\037\001\t\t\t\t\t\t\t\t\001\t\t\t\t\t\t\t\t") at strnatcmp.c:66
---Type <return> to continue, or q <return> to quit---
```

As `#25000` is still the same message (and it's not the last one) it seems that it endlessly calls itself in the last line.

**Comment by flyingmutant** [comment]
Here is the version with proper recursion termination — https://github.com/flyingmutant/cmus/commit/5d934b2ad2b4e9da3a5136df80b1e4f5cf376fa5. I've also used `strnatcmp` for the browser view. I like it more because the code is still much simpler, yet does the same thing (I hope :-)

Can you please test if it does work for you as well?

**Comment by jayay** [comment]
Yes, that looks good to me! Next step is to write `u_strncasecoll` then :grin:

Did you try out filenames with cyrillic and chinese characters? I hope the order is still okay in that case. I wish I could help, but I really don't have a clue in both of them.

**Comment by flyingmutant** [comment]
Merged in fc3490d6e251cd38be096f7d56047ca73ac73361!

**Comment by jayay** [comment]
Wooo! Thank you!
